### ISSUE TITLE
[8章-2]BERT用LayerNormalizationについて

### ISSUE THREAD

**Issue by ryota0051** [YutaroOgawa/pytorch_advanced#101]
お世話になっております。

Bert用のLayerNormalizationを実装する部分(https://github.com/YutaroOgawa/pytorch_advanced/blob/master/8_nlp_sentiment_bert/8-2-3_bert_base.ipynb の class BertLayerNormを実装する部分のことです。)について以下の疑問があります。

* 質問内容

LayerNormalizationは、特徴量ごとに平均と分散を計算しデータの平均と分散をそれぞれ0および1にするというアルゴリズムだと解釈しています。なので、単語の特徴量ベクトルだけではなく、単語数に関しても平均と分散を計算する必要があると感じました(コード的には、下記になると思います。)。なぜ、書籍における実装は、単語ベクトルのみの平均と分散を計算しているのでしょうか？

```Python
class BertLayerNorm(nn.Module):
    """LayerNormalization層 """

    def __init__(self, hidden_size, eps=1e-12):
        super(BertLayerNorm, self).__init__()
        self.gamma = nn.Parameter(torch.ones(hidden_size))  # weightのこと
        self.beta = nn.Parameter(torch.zeros(hidden_size))  # biasのこと
        self.variance_epsilon = eps

    def forward(self, x):
        # meanのdimに(単語数の軸インデックス, 特徴量ベクトルの軸インデックス)を与えるように修正しています。)
        u = x.mean(dim=(1, 2), keepdim=True)
        s = (x - u).pow(2).mean(dim=(1, 2), keepdim=True) 
        x = (x - u) / torch.sqrt(s + self.variance_epsilon)
        return self.gamma * x + self.beta
```

*参照した記事

1. https://www.tensorflow.org/addons/tutorials/layers_normalizations

     tensorflowのlayer normalizationsの説明に関する記事で、layer normalizationsがどのような動作をしているか確認するために参照しました。この記事から、バッチの次元以外の平均を取る必要があるのでは？と疑問に思いました。

2. https://pytorch.org/docs/master/generated/torch.mean.html

    torch.meanに関する記事で、dimの引数に複数の軸を与えられることを確認するために参照しました。

以上、よろしくお願いいたします。

**Comment by ryota0051** [comment]
追加で調べていたのですが、単語数の次元は画像でいうchannelとして扱うべきではない(どちらかというと、バッチに近いという印象です。)ので、dim = -1としているという認識でいいのでしょうか?

**Comment by YutaroOgawa** [comment]
@ryota0051 さま

ご質問、ありがとうございます。
まず、こちらのclass BertLayerNormはHuggingFaceさんのBERT実装をそのままお借りしてきた部分になります。

本書が出版された頃は、彼はPyTorchとTensorFlowで、
LayerNormlization層の挙動が異なるため、実装し直していました。

最新版では、HuggingFaceさんも、

BertLayerNorm = torch.nn.LayerNorm

として、PyTorchのLayerNormを使用しています。

https://github.com/huggingface/transformers/blob/614fef1691edb806de976756d4948ecbcd0c0ca3/src/transformers/modeling_bert.py


で、このレイヤーノーマライゼーションですが、以下となります。

https://pytorch.org/docs/master/generated/torch.nn.LayerNorm.html


レイヤーノーマライゼーションの説明ですが、以下のブログが分かりやすいです。
https://blog.albert2005.co.jp/2018/09/05/group_normalization/

![groupnorm-gn-zu-1024x256](https://user-images.githubusercontent.com/38151872/88470068-eecdca80-cf32-11ea-8620-22aa3f78a628.png)


レイヤーノーマライゼーションは、画像であれば、チャネル方向に正規化をかけます（その該当範囲が平均0標準偏差1になる変換を求める）

今回のテキストデータ、BERTの場合は、単語特徴量のみをレイヤーノーマライゼーションの対象にしています。

私は元のコードも、
u = x.mean(-1, keepdim=True)
s = (x - u).pow(2).mean(-1, keepdim=True)

も、いただいたコード例

u = x.mean(dim=(1, 2), keepdim=True)
s = (x - u).pow(2).mean(dim=(1, 2), keepdim=True) 

も、ほぼほぼ似た同じ動作をするような気がしています。

BERTはただ、「各単語の特徴量768次元の各次元は平均0、標準偏差1」とすると決めただけであって。

ただ、思想が上記の2つのコードでは異なり、

元のコードの場合は
「各単語の特徴量768次元の各次元は平均0、標準偏差1」とする
です。

一方で、いただいた例では
「各単語の特徴量768次元の各次元は、512単語方向も考慮し平均0、標準偏差1」とする
になってしまいます。

つまり、平均に単語方向を入れると、
「単語方向も考慮して、正規化するので、1単語レベルではなく、入ってきている単語全部も考慮して、各単語の768次元の特徴量が正規化できれば良いや！」

という考え方になります。

これが誤りだというわけではなく、ただ、BERTはそういう思想ではなく1単語レベルで正規化したい、というものになります。

以上、かなり、混みいった話になりますが、解説になります。

私も非常に考えさせられる投稿でした。
理解しづらい文章もあるかと思いますので、気になった点については、引き続き、ご質問ください。
素晴らしいIssueをありがとうございます。

**Comment by ryota0051** [comment]
ご返事ありがとうございます。

おかげさまで概ね理解することができましたが2つ疑問があるので質問させていただきます。

1. PyTorchとTensorFlowで、LayerNormlization層の挙動が異なるということですが、どのような点が異なっていたのか教えていただけると嬉しいです。

2. 単語方向も含めて正規化する方法で実装してみましたところ、bankの類似度を確かめる部分において、挙動が異なっていました(具体的には、bertのAttentionにおける最終層出力ベクトルのコサイン類似度がほぼ同じ値になります。)。これは、学習時と推論時で異なる正規化方法を採用しているためという認識でよろしいでしょうか(画像分野でいうと、学習時は、データを255で割り算する前処理をしていたが、推論時は、-1〜1の間の値に収める前処理を行ってしまったという状況かと思っています。)?

以上、お手数ですが、よろしくお願いいたします。

**Comment by YutaroOgawa** [comment]
@ryota0051 さま

ありがとうございます。

[1] 
>PyTorchとTensorFlowで、LayerNormlization層の挙動が異なるということですが、どのような点が異なっていたのか教えていただけると嬉しいです。

以下リンクは、PyTorchのバージョンが0.4のときのLayernormの式です。
https://pytorch.org/docs/0.4.0/nn.html?highlight=layernorm#torch.nn.LayerNorm

そして、v1.0以降のLayernormの式です（TensorFlowと同じ）。
https://pytorch.org/docs/master/generated/torch.nn.LayerNorm.html

0除算を避けるためのεが入るのが、v0.4のころはルートの外にεがいたのが、
TensorFlowに合わせて、ルートの中に変わりました。
（すなわち、TensorFlow版にPyTorch側が変更した）

これを受けて、HuggingFaceさんも、LayerNormはPyTorchの標準を今は使用しています。
（なお本書はPyTorchのバージョンが0.4から1.0の過渡期で書いたので、LayerNormalizationを自分たちで定義し直しています）

[2] 状況が独特すぎるので、なんとも分かりませんが、まず同じLayernormlizationの正規化手法を実装しているなら、学習も推論も挙動は同じになるはずです。
そして、単純に学習と推論で前処理が異なると、それは学習と推論で傾向も変わってしまいます。
ただ、おそらく今回は、　@ryota0051 さんが、 「単語方向まで考えたノーマライゼーション」を試していますが、
学習済みデータはそもそも「単語方向は考えず、各単語の特徴量でノーマライゼーション」したものを転移学習しているので、「ベースの学習済みモデル」と「やりたいlayernormalization」が不一致な点が問題の原因な気がします。

どうぞよろしくお願い致します。
